File: LeetCode - Algorithm/Linked List/19.py
# 19. Remove Nth Node From End of List
# https://leetcode.com/problems/remove-nth-node-from-end-of-list/

from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if not head:
            return head

        prev = ListNode(-1)
        prev.next = head

        target = head
        fast = head

        for _ in range(n):
            fast = fast.next

        if not fast:
            return target.next

        while fast:
            prev = prev.next
            target = target.next
            fast = fast.next

        prev.next = target.next

        return head

File: LeetCode - Algorithm/Linked List/19.js
// 19. Remove Nth Node From End of List
// https://leetcode.com/problems/remove-nth-node-from-end-of-list/

function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
// time complexity: O(n)
// space complexity: O(1)
var removeNthFromEnd = function (head, n) {
  let nodeBeforeSlow = null;
  let slow = head;
  let fast = head;

  for (let i = 0; i < n; i++) {
    fast = fast.next;
  }

  while (fast) {
    fast = fast.next;
    nodeBeforeSlow = slow;
    slow = slow.next;
  }

  if (!nodeBeforeSlow) return slow.next;

  nodeBeforeSlow.next = slow.next;

  return head;
};
